### PR TITLE
8327946: containers/docker/TestJFREvents.java fails when host kernel config vm.swappiness=0 after JDK-8325139

### PR DESCRIPTION
**Pull request by sendaoYan**
Hi,

According to the [docker document](https://docs.docker.com/config/containers/resource_constraints/#--memory-swappiness-details), the default value of --memory-swappiness is inherited from the host machine. So, when the the kernel config vm.swappiness=0 on the host machine, this testcase will fail, because of docker container can not use swap memory, the deafult value of --memory-swappiness is 0.

When the host kernel config "vm.swappiness = 0", In order to run this testcase passed , there are three methods:

1. change `.shouldContain("totalSize = " + expectedTotalValue)` to `.shouldContain("totalSize = "`, which ignored the `expectedTotalValue`, because the `expectedTotalValue` could be 0(swap memroy is disable when --memory-swappiness=0) or could be 104857600(300MB-200MB=100MB), it depends on the host machine config `vm.swappiness`
2. Change the default `--memory-swappiness` 0 to non-zero, such as 60.
3. Change the host kernel config `vm.swappiness=0` to `vm.swappiness=60`. I think it's not a good idea.

Maybe the 2rd method seems more resonable.

Thanks,
-sendao

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327946](https://bugs.openjdk.org/browse/JDK-8327946): containers/docker/TestJFREvents.java fails when host kernel config vm.swappiness=0 after JDK-8325139 (**Bug** - P4)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18225/head:pull/18225` \
`$ git checkout pull/18225`

Update a local copy of the PR: \
`$ git checkout pull/18225` \
`$ git pull https://git.openjdk.org/jdk.git pull/18225/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18225`

View PR using the GUI difftool: \
`$ git pr show -t 18225`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18225.diff">https://git.openjdk.org/jdk/pull/18225.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18225#issuecomment-1991111164)

### CODE REVIEW

**Comment by sendaoYan** [comment]
Fix the testcase bug, the risk is low. Can anyone reivew this PR, thanks.

**Comment by sendaoYan** [comment]
Fix the testcase bug, the risk is low. Can anyone reivew this PR, thanks.

**Review by jerboaa — CHANGES_REQUESTED** [review]
`--memory-swappiness` is cgv1 specific and needs to be handled for the cgv2 case.

File: test/hotspot/jtreg/containers/docker/TestJFREvents.java [hunk 1]
```diff
@@ -221,6 +221,8 @@ private static void testSwapMemory(String memValueToSet, String swapValueToSet,
                                       commonDockerOpts()
                                       .addDockerOpts("--memory=" + memValueToSet)
                                       .addDockerOpts("--memory-swap=" + swapValueToSet)
+                                      //The default memory-swappiness vaule is inherited from the host machine, which maybe 0
+                                      .addDockerOpts("--memory-swappiness=60")
```
[the remark this answers is not in the document]
Comment (sendaoYan) [reply test/hotspot/jtreg/containers/docker/TestJFREvents.java]: Thanks for your review. The space after `//` has been added.

I can't reproduce the "OCI runtime error" failure on mine ubuntu22 environment.
It seems that ubuntu22 use cgroups v2 by default.
image

Can you show your host machine enviroment information, so I can reproduce the same failure. After that I will try to find a solution with cgroupv2.

[the remark this answers is not in the document]
Comment (jerboaa) [reply test/hotspot/jtreg/containers/docker/TestJFREvents.java]: It seems to be podman runtime specific. `crun` fails, `runc` doesn't seem to be. Either way, the corresponding interface file, `memory.swappiness` doesn't exist for cgroup v2. Try `podman run --runtime /usr/bin/crun --rm -ti --memory-swappiness=60 fedora:39`  provided the `crun` runtime is installed in `/usr/bin`.

**Review by jerboaa — COMMENTED** [review]
I don't think we need `Whitebox`. A simpler check would be to use the `Metrics` class. See for example `test/hotspot/jtreg/containers/docker/TestMemoryWithCgroupV1.java`.

**Comment by sendaoYan** [comment]
> I don't think we need `Whitebox`. A simpler check would be to use the `Metrics` class. See for example `test/hotspot/jtreg/containers/docker/TestMemoryWithCgroupV1.java`.

Thanks your advice. The `Whitebox` has been removed, and Use `Metrics`class to get the cgroup version information.

**Review by jerboaa — COMMENTED** [review]
Better. I don't think we need to duplicate that much of the test. All we'd need to do is to add the `--memory-swappiness` option only if we are on cg v1:

```
                                          opts.addDockerOpts("--memory=" + memValueToSet)
                                                  .addDockerOpts("--memory-swap=" + swapValueToSet)
                                                  .addClassOptions("jdk.SwapSpace"));
                                          if (isCgroupV1) {
                                              // With Cgroupv1, The default memory-swappiness vaule is inherited from the host machine, which maybe 0
                                              opts.addDockerOpts("--memory-swappiness=60");
                                          }
                                          out = DockerTestUtils.dockerRunJava(opts);
                                          ...
```

**Comment by sendaoYan** [comment]
> Better. I don't think we need to duplicate that much of the test. All we'd need to do is to add the `--memory-swappiness` option only if we are on cg v1:
> 
> ```
>                                           opts.addDockerOpts("--memory=" + memValueToSet)
>                                                   .addDockerOpts("--memory-swap=" + swapValueToSet)
>                                                   .addClassOptions("jdk.SwapSpace"));
>                                           if (isCgroupV1) {
>                                               // With Cgroupv1, The default memory-swappiness vaule is inherited from the host machine, which maybe 0
>                                               opts.addDockerOpts("--memory-swappiness=60");
>                                           }
>                                           out = DockerTestUtils.dockerRunJava(opts);
>                                           ...
> ```

Thank you very much for your very professional opinions. The modification of the review opinions has been completed. In addition, an extra space at the beginning of the line was deleted.

**Review by jerboaa — APPROVED** [review]
Looks OK to me.

**Comment by jerboaa** [comment]
If you merge latest master, GHA should be clean too.

**Comment by sendaoYan** [comment]
> If you merge latest master, GHA should be clean too.



> If you merge latest master, GHA should be clean too.

The GHA show java/lang/String/StringRepeat.java#id1 fails on linux x86. It's a testcase bug which has been fixed in [8328524](https://github.com/openjdk/jdk/pull/18380)

**Comment by sendaoYan** [comment]
> Looks OK to me.

Thanks for the review.

**Comment by sendaoYan** [comment]
> /sponsor

Thanks.

**Pull request closed without merging**